modules/porous_flow/src/materials/PorousFlow2PhaseHysPS.C

```diff
@@ -0,0 +1,120 @@
+                                         : coupledValue("phase0_porepressure")),
+    _phase0_gradp_qp(coupledGradient("phase0_porepressure")),
+    _phase0_porepressure_varnum(coupled("phase0_porepressure")),
+    _pvar(_dictator.isPorousFlowVariable(_phase0_porepressure_varnum)
+              ? _dictator.porousFlowVariableNum(_phase0_porepressure_varnum)
+              : 0),
+
+    _phase1_saturation(_nodal_material ? coupledDofValues("phase1_saturation")
+                                       : coupledValue("phase1_saturation")),
+    _phase1_grads_qp(coupledGradient("phase1_saturation")),
+    _phase1_saturation_varnum(coupled("phase1_saturation")),
+    _svar(_dictator.isPorousFlowVariable(_phase1_saturation_varnum)
+              ? _dictator.porousFlowVariableNum(_phase1_saturation_varnum)
+              : 0)
+{
+  if (_num_phases != 2)
+    mooseError("The Dictator announces that the number of phases is ",
+               _dictator.numPhases(),
+               " whereas PorousFlow2PhaseHysPS can only be used for 2-phase simulation.  The "
+               "Dictator is always watching.");
+}
+
+void
+PorousFlow2PhaseHysPS::initQpStatefulProperties()
+{
+  PorousFlowHystereticCapillaryPressure::initQpStatefulProperties();
+  buildQpPPSS();
+}
+
+void
+PorousFlow2PhaseHysPS::computeQpProperties()
+{
+  // size stuff correctly and prepare the derivative matrices with zeroes
+  PorousFlowHystereticCapillaryPressure::computeQpProperties();
+
+  buildQpPPSS();
+  const Real dpc = dcapillaryPressureQp(1.0 - _phase1_saturation[_qp]); // d(Pc)/d(S0)
+
+  if (!_nodal_material)
+  {
+    (*_grads_qp)[_qp][0] = -_phase1_grads_qp[_qp];
+    (*_grads_qp)[_qp][1] = _phase1_grads_qp[_qp];
+    (*_gradp_qp)[_qp][0] = _phase0_gradp_qp[_qp];
+    (*_gradp_qp)[_qp][1] = _phase0_gradp_qp[_qp] - dpc * (*_grads_qp)[_qp][1];
+  }
+
+  // _porepressure depends on _phase0_porepressure, and its derivative is 1
+  if (_dictator.isPorousFlowVariable(_phase0_porepressure_varnum))
+  {
+    // _phase0_porepressure is a PorousFlow variable
+    for (unsigned phase = 0; phase < _num_phases; ++phase)
+    {
+      _dporepressure_dvar[_qp][phase][_pvar] = 1.0;
+      if (!_nodal_material)
+        (*_dgradp_qp_dgradv)[_qp][phase][_pvar] = 1.0;
+    }
+  }
+
+  // _saturation is only dependent on _phase1_saturation, and its derivative is +/- 1
+  if (_dictator.isPorousFlowVariable(_phase1_saturation_varnum))
+  {
+    // _phase1_saturation is a PorousFlow variable
+    // _phase1_porepressure depends on saturation through the capillary pressure function
+    _dsaturation_dvar[_qp][0][_svar] = -1.0;
+    _dsaturation_dvar[_qp][1][_svar] = 1.0;
+    _dporepressure_dvar[_qp][1][_svar] = -dpc;
+
+    if (!_nodal_material)
+    {
+      (*_dgrads_qp_dgradv)[_qp][0][_svar] = -1.0;
+      (*_dgrads_qp_dgradv)[_qp][1][_svar] = 1.0;
+      const Real d2pc_qp = d2capillaryPressureQp(1.0 - _phase1_saturation[_qp]); // d^22(Pc)/dS0^2
```

Comment:
Super minor but only the second derivative not the 22nd!

Reply:
Heheheh.  As i've said before, your eyes are amazing.